### DISCUSSION TITLE
Octo => open-source replacement for Obsidian?

### DISCUSSION THREAD

**Discussion by PhilLovesToCode** [davidmyersdev/octo#200] [Ideas]
Have you thought about morphing Octo into an open-source alternative to Obsidian?  Obsidian has gained an enormous following despite the fact that it is closed source.  I suspect that many developers would join in and become contributors should you move in that direction.

**Comment by davidmyersdev** [comment]
Hey @PhilLovesToCode! 👋

Thanks so much for taking the time to start this conversation. I think it makes a lot of sense to offer a path for users to migrate from Obsidian to Octo, and the fact that Obsidian is closed source is _one_ of the main reasons I opted not to use it myself.

Which features would make Octo a viable alternative to Obsidian for you (and others)?

**Reply by PhilLovesToCode** [reply]
It's challenging because Obsidian is very mature and feature-rich, made more so by it's plugin ecosystem.  So you need to think in terms of what would be a minimum viable product (MVP). IMO the greatest single attraction of Obsidian is the fact that it stores its documents in a normal folder/directory tree in the file system, which they call the "vault", and which they clearly show in a tree view on the left-hand sidebar.  Also that it is available as a desktop app on Mac, Windows, Linux, and iOS. (It's probably developed using Electron, but for a greenfield project, Tauri might be a better choice.)  These two features leave the user feeling very much in control of their data.  Obsidian's monetization model is their sync feature which works extremely well.  It syncs my vault across all of my different platforms, which for me include all mentioned above.

Having said all this, it's only one person's opinion and POV.  I hope others will join the discussion with their ideas about what should be in a MVP.

**Reply by davidmyersdev** [reply]
Thanks so much for sharing all this feedback! I'll do my best to provide my own perspective on how I think Octo compares and what I have in mind for each of these points you brought up. None of this is set in stone though, and feedback like this absolutely helps me prioritize which features should be done sooner rather than later.

> It's challenging because Obsidian is very mature and feature-rich, made more so by it's plugin ecosystem.

This is a fair point. I have started working on a plugin system for Octo as well, but the internal interfaces are still changing pretty frequently. Beyond stability, security is the most important factor since the app runs in the browser, so the plugin system will have to be built with a little more care than typical electron-based implementations. At the moment, I'm looking to implement plugin sandboxing via web workers, web assembly modules, and possibly iframes (for plugin UIs). The main thing I like to keep in mind is making sure this all can run both natively and in the browser. Browsers and native apps are quickly approaching feature parity, and many of the features we typically think can only be done on a native device are actually possible on the web now.

> IMO the greatest single attraction of Obsidian is the fact that it stores its documents in a normal folder/directory tree in the file system, which they call the "vault", and which they clearly show in a tree view on the left-hand sidebar.

The ability to work on local files (one-by-one) is already available in Octo as an "experimental" setting, but that will likely be exposed as a mainstream feature once folder support (essentially Obsidian's "Vault" feature) is finished. This is all possible without native apps via the File System Access API, but there are some limitations which is why "native shells" are on the roadmap. That said, one goal of Octo was to break free of the hierarchical structure imposed by folders and instead use tags to map document relationships. I think having the ability to use both will be good though.

> Also that it is available as a desktop app on Mac, Windows, Linux, and iOS.

As I mentioned above, I have plans for "native shells" (wrappers around the web app), but Octo currently works on any device since it's a responsive, offline-first progressive web app. Some devices (iOS) don't have full PWA support, but the important features (offline support, client-side encryption, etc) still work as-is.

> These two features leave the user feeling very much in control of their data.

Data ownership and privacy are huge tenets of mine, so I absolutely understand this desire. With Octo, documents are stored on your local filesystem (using IndexedDB) first and only backed up/synced if you have an account, are signed in, and are online. Even then, you can enable client-side encryption to prevent those docs from traveling anywhere in plain text. I still need to add the ability to _disable_ sync for some or all documents even when signed in, but that's also on the roadmap. Along this line, you can also export your entire collection of documents as JSON and (in the near future) a collection of markdown files.

> Obsidian's monetization model is their sync feature which works extremely well. It syncs my vault across all of my different platforms, which for me include all mentioned above.

Octo's monetization model is similar, although I'm primarily going to be focusing on teams and enterprise features for the paid plan (real-time collaboration, publishing collections as "wikis", fine-grained multi-user permissions, etc). The free plan includes unlimited syncing between devices and read-only doc sharing.
